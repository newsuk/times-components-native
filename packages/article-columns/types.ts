import { ParagraphContent } from "./domain-types";

export type Line = {
  text: string;
};
export type TextLayoutEvent = { lines: Line[] };
export type ColumnContents = ParagraphContent[];
export interface PageColumn {
  byline?: {
    height: number;
  };
  image?: {
    height: number;
    margin: number;
  };
  contents: ColumnContents;
  contentHeight: number;
}

export interface ArticleMeasurements {
  contents: {
    lines: { [key: string]: Line[] };
    heights: { [key: string]: number };
  };
  bylineHeight: number | null;
  bylineMargin: number | null;
}

export interface ColumnParameters {
  columnWidth: number;
  columnHeight: number;
  columnCount: number;
  columnLineHeight: number;
}

export interface FontScale {
  fontSize: number;
  lineHeight: number;
}
