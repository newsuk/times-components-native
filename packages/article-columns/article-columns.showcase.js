import React from "react";
import { ArticleColumns } from "./article-columns";
import article from "./fixtures/article.json";
export default {
  children: [
    {
      component: () => (
        <ArticleColumns
          articleContents={article.data.article.content}
          containerHeight={500}
          containerWidth={500}
          lineHeight={18}
          bylines={article.data.article.bylines}
          style={{ lineHeight: 18 }}
          columnCount={2}
        />
      ),
      name: "Columns",
      type: "story",
    },
  ],
  name: "Composed/Article Columns",
};
