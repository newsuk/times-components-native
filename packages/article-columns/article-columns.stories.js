import { showcaseConverter } from "@times-components-native/storybook";
import showcase from "./article-columns.showcase";

showcaseConverter(module, showcase);
