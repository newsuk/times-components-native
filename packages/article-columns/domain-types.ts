export type Bylines = {
  byline: Byline[];
};

type Byline = {
  name: string;
  attributes: any;
  children: BylineChildren[];
};

type BylineChildren = {
  name: string;
  attributes: {
    value: string;
  };
  children: any[];
};

interface TextContentChild {
  name: "text";
  children: ParagraphContentChild[];
  attributes: {
    value: string;
  };
}

interface LinkContentChild {
  name: "link";
  attributes?: any;
  children: ParagraphContentChild[];
}

interface ItalicContentChild {
  name: "italic";
  attributes?: any;
  children: ParagraphContentChild[];
}

interface BoldContentChild {
  name: "bold";
  attributes?: any;
  children: ParagraphContentChild[];
}

interface BreakContentChild {
  name: "break";
  attributes?: any;
  children: ParagraphContentChild[];
}

interface InvisibleContentChild {
  name: "invisible";
  children: ParagraphContentChild[];
  attributes: {
    value: string;
  };
}

export type ParagraphContentChild =
  | TextContentChild
  | ItalicContentChild
  | LinkContentChild
  | BreakContentChild
  | BoldContentChild
  | InvisibleContentChild;

export interface ContentMetadata {
  id?: string;
  split?: boolean;
}

export interface ParagraphContent extends ContentMetadata {
  name: "paragraph";
  children: ParagraphContentChild[];
  attributes?: {
    tab?: boolean;
  };
}

interface InteractiveContent extends ContentMetadata {
  name: "interactive";
  children: [];
}

export interface ImageContent extends ContentMetadata {
  name: "image";
  attributes: {
    display: string;
    caption: string;
    credits: string;
    url: string;
    relativeHorizontalOffset: number;
    relativeVerticalOffset: number;
    relativeWidth: number;
    relativeHeight: number;
  };
  children: [];
}

interface AdContent extends ContentMetadata {
  name: "ad";
  children: [];
}

export type ArticleContent =
  | ParagraphContent
  | InteractiveContent
  | ImageContent
  | AdContent;
