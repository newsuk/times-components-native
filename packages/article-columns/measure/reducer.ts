import { Reducer } from "react";
import {
  ArticleMeasurements,
  Line,
} from "@times-components-native/article-columns/types";

export const initialState: ArticleMeasurements = {
  contents: {
    lines: {},
    heights: {},
  },
  bylineHeight: null,
  bylineMargin: null,
};

interface SetContentHeightAction {
  type: "SET_CONTENT_HEIGHT";
  id: string;
  payload: number;
}

interface SetContentLinesAction {
  type: "SET_CONTENT_LINES";
  id: string;
  payload: Line[];
}

interface SetBylineAction {
  type: "SET_BYLINE_HEIGHT";
  height: number;
  margin: number;
}

export type Action =
  | SetContentHeightAction
  | SetContentLinesAction
  | SetBylineAction;

export const reducer: Reducer<ArticleMeasurements, Action> = (
  state,
  action,
) => {
  switch (action.type) {
    case "SET_CONTENT_HEIGHT":
      return {
        ...state,
        contents: {
          ...state.contents,
          heights: { ...state.contents.heights, [action.id]: action.payload },
        },
      };
    case "SET_CONTENT_LINES":
      return {
        ...state,
        contents: {
          ...state.contents,
          lines: { ...state.contents.lines, [action.id]: action.payload },
        },
      };
    case "SET_BYLINE_HEIGHT":
      return {
        ...state,
        bylineHeight: action.height,
        bylineMargin: action.margin,
      };
    default:
      return state;
  }
};
