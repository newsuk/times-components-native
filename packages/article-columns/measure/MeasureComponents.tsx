import renderTrees from "@times-components-native/markup-forest";
import React, { memo } from "react";
import { TextStyle, View } from "react-native";
import {
  Bylines,
  ParagraphContent,
} from "@times-components-native/article-columns/domain-types";
import { getRenderers } from "@times-components-native/front-page/front-renderer";

import { useMeasurementDispatchContext } from "./MeasurementDispatchContext";
import { ColumnParameters } from "@times-components-native/article-columns/types";
import { FrontPageByline } from "@times-components-native/front-page/front-page-byline";

interface Props {
  content: ParagraphContent;
  style: TextStyle;
}
export const MeasureContent: React.FC<Props> = memo(({ content, style }) => {
  const dispatch = useMeasurementDispatchContext();

  return (
    <View
      onLayout={(event) => {
        const height = Math.floor(event.nativeEvent.layout.height);
        dispatch({
          type: "SET_CONTENT_HEIGHT",
          id: content.id!,
          payload: height,
        });
      }}
    >
      {renderTrees(
        [content],
        getRenderers({
          renderOptions: style,
          onParagraphTextLayout: (event: any) => {
            const lines = event.nativeEvent.lines;
            dispatch({
              type: "SET_CONTENT_LINES",
              id: content.id!,
              payload: lines,
            });
          },
        }),
      )}
    </View>
  );
});

export const calculateDynamicSpacing = (height: number, lineHeight: number) => {
  const padding = Math.ceil(height / lineHeight) * lineHeight - height;
  return padding >= 7 ? padding : padding + lineHeight;
};

export const MeasureByline: React.FC<{
  bylines: Bylines;
  columnParameters: ColumnParameters;
}> = memo(({ bylines, columnParameters }) => {
  const dispatch = useMeasurementDispatchContext();

  return (
    <View
      style={{ width: columnParameters.columnWidth }}
      onLayout={(event) => {
        const height = event.nativeEvent.layout.height;
        dispatch({
          type: "SET_BYLINE_HEIGHT",
          height,
          margin: calculateDynamicSpacing(
            height,
            columnParameters.columnLineHeight,
          ),
        });
      }}
    >
      <FrontPageByline
        byline={bylines}
        withKeyline={true}
        containerStyle={{ marginBottom: 0 }}
      />
    </View>
  );
});
