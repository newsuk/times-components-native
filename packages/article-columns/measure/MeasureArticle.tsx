import React, { useReducer } from "react";
import { ScrollView, StyleSheet, TextStyle, View } from "react-native";

import { ColumnParameters, Line, ArticleMeasurements } from "../types";

import { MeasureByline, MeasureContent } from "./MeasureComponents";
import { MeasurementDispatch } from "./MeasurementDispatchContext";
import { initialState, reducer } from "./reducer";
import { ParagraphContent } from "../domain-types";
import { BylineInput } from "@times-components-native/fixture-generator/src/types";

interface Props {
  articleContents: ParagraphContent[];
  bylines: BylineInput[];
  columnParameters: ColumnParameters;
  renderMeasuredContents: (articleMeasurements: ArticleMeasurements) => any;
  style: TextStyle;
}

const allContentMeasured = (
  contentToMeasure: ParagraphContent[],
  heights: { [key: string]: number },
  lines: { [key: string]: Line[] },
) => {
  return (
    contentToMeasure.filter((c) => heights[c.id!] && lines[c.id!]).length ===
    contentToMeasure.length
  );
};

export const InnerMeasureArticle: React.FC<
  Props & { measurementState: ArticleMeasurements }
> = ({
  articleContents,
  bylines,
  columnParameters,
  renderMeasuredContents,
  measurementState,
  style,
}) => {
  const {
    contents: { lines, heights },
  } = measurementState;
  if (
    allContentMeasured(articleContents, heights, lines) &&
    measurementState.bylineHeight !== null
  ) {
    return renderMeasuredContents(measurementState);
  }

  return (
    <ScrollView style={styles.renderOffscreen}>
      <MeasureByline columnParameters={columnParameters} bylines={bylines} />
      <View style={{ width: columnParameters.columnWidth }}>
        {articleContents.map((content) => (
          <MeasureContent
            key={`ContentMeasuringView:${content.id}`}
            style={style}
            content={content}
          />
        ))}
      </View>
    </ScrollView>
  );
};

/* istanbul ignore next */
export const MeasureArticle: React.FC<Props> = (props) => {
  const [measurementState, measurementDispatch] = useReducer(
    reducer,
    initialState,
  );

  return (
    <MeasurementDispatch.Provider value={measurementDispatch}>
      <InnerMeasureArticle {...props} measurementState={measurementState} />
    </MeasurementDispatch.Provider>
  );
};

const styles = StyleSheet.create({
  renderOffscreen: {
    transform: [{ translateX: -1000 }],
  },
});
