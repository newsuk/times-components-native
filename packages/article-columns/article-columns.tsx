import React from "react";
import { MeasureArticle } from "./measure/MeasureArticle";
import {
  ArticleContent,
  Bylines,
  ParagraphContent,
} from "@times-components-native/article-columns/domain-types";
import { chunkContentIntoColumns } from "@times-components-native/article-columns/utils/chunkContentIntoColumns";
import {
  Columns,
  SingleColumn,
} from "@times-components-native/article-columns/render/Columns";
import { TextStyle } from "react-native";
import { calculateColumnWidth } from "@times-components-native/article-columns/utils/calculateColumnWidth";

interface Props {
  articleContents: ArticleContent[];
  bylines: Bylines;
  containerWidth: number;
  containerHeight: number;
  lineHeight: number;
  columnCount: number;
  style: TextStyle;
}

const assignWithId = (c: ParagraphContent, idx: number): ParagraphContent => {
  c.id = `${idx}`;
  return c;
};

export const ArticleColumns: React.FC<Props> = ({
  articleContents,
  bylines,
  containerHeight,
  containerWidth,
  columnCount,
  lineHeight = 18,
  style,
}) => {
  const paragraphs = articleContents
    .filter((c): c is ParagraphContent => c.name === "paragraph")
    .map(assignWithId);

  const columnWidth = calculateColumnWidth({
    columnCount,
    columnGap: 10,
    containerWidth,
  });
  const columnParameters = {
    columnWidth,
    columnHeight: containerHeight,
    columnCount,
    columnLineHeight: lineHeight,
  };

  return (
    <MeasureArticle
      bylines={bylines}
      articleContents={paragraphs}
      columnParameters={columnParameters}
      style={style}
      renderMeasuredContents={(articleMeasurements) => {
        const articleColumns = chunkContentIntoColumns(
          paragraphs,
          articleMeasurements,
          columnParameters,
        ).slice(0, columnCount);
        return (
          <Columns>
            {articleColumns.map((articleColumn, index) => (
              <SingleColumn
                key={index}
                style={style}
                columnParameters={columnParameters}
                column={articleColumn}
              />
            ))}
          </Columns>
        );
      }}
    />
  );
};
