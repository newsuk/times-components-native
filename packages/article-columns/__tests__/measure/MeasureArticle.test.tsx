import React from "react";
import TestRenderer from "react-test-renderer";

import { InnerMeasureArticle as MeasureArticle } from "@times-components-native/article-columns/measure/MeasureArticle";
import { initialState } from "@times-components-native/article-columns/measure/reducer";
import {
  ArticleContent,
  ParagraphContent,
} from "@times-components-native/article-columns/domain-types";
import { ColumnParameters } from "@times-components-native/article-columns/types";
import MockMarkup from "@times-components-native/fixture-generator/src/mock-markup";

jest.mock(
  "@times-components-native/article-columns/measure/MeasureComponents",
  () => ({
    MeasureByline: "MeasureByline",
    MeasureHeadline: "MeasureHeadline",
    MeasureContent: "MeasureContent",
  }),
);

const createParagraphWithText = (
  text: string,
  testData: Partial<ParagraphContent> = {},
): ParagraphContent => ({
  id: "some-paragraph-id",
  name: "paragraph",
  children: [{ name: "text", children: [], attributes: { value: text } }],
  ...testData,
});

const paragraphA: ArticleContent = createParagraphWithText("line1", {
  id: "p1",
});

// eslint-disable-next-line @typescript-eslint/no-unused-vars
const CaptureContents = ({ articleMeasurements }: any) => null;

const columnParameters: ColumnParameters = {
  columnHeight: 2000,
  columnWidth: 110,
  columnCount: 3,
  columnLineHeight: 20,
};

const bylines = new MockMarkup().addBylines().get();
const style = {
  fontSize: 14,
  lineHeight: 18,
};
describe("MeasureArticle", () => {
  it("renders everything to be measured", () => {
    const renderer = TestRenderer.create(
      <MeasureArticle
        bylines={bylines}
        columnParameters={columnParameters}
        style={style}
        articleContents={[paragraphA]}
        renderMeasuredContents={() => null}
        measurementState={initialState}
      />,
    );

    expect(renderer.toJSON()).toMatchSnapshot();
  });

  it("outputs rendered content once everything has been measured", () => {
    const renderer = TestRenderer.create(
      <MeasureArticle
        bylines={bylines}
        columnParameters={columnParameters}
        style={style}
        articleContents={[paragraphA]}
        renderMeasuredContents={(articleMeasurements) => (
          <CaptureContents articleMeasurements={articleMeasurements} />
        )}
        measurementState={{
          contents: {
            lines: { p1: [{ text: "line1" }] },
            heights: { p1: 20 },
          },
          bylineHeight: 20,
          bylineMargin: 0,
        }}
      />,
    );

    expect(renderer.toJSON()).toMatchSnapshot();

    const articleMeasurements = renderer.root.findByType(CaptureContents).props[
      "articleMeasurements"
    ];
    expect(articleMeasurements).toEqual({
      bylineHeight: 20,
      contents: {
        lines: {
          p1: [{ text: "line1" }],
        },
        heights: {
          p1: 20,
        },
      },
    });
  });
});
