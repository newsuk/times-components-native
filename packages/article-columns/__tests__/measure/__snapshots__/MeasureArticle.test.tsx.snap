// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MeasureArticle outputs rendered content once everything has been measured 1`] = `null`;

exports[`MeasureArticle renders everything to be measured 1`] = `
<RCTScrollView
  style={
    Object {
      "transform": Array [
        Object {
          "translateX": -1000,
        },
      ],
    }
  }
>
  <View>
    <MeasureByline
      bylines={
        Array [
          Object {
            "byline": Array [
              Object {
                "attributes": Object {
                  "slug": "richard-lloyd-parry",
                },
                "children": Array [
                  Object {
                    "attributes": Object {
                      "value": "Richard Lloyd Parry",
                    },
                    "children": Array [],
                    "name": "text",
                  },
                ],
                "name": "author",
              },
            ],
            "image": null,
          },
          Object {
            "byline": Array [
              Object {
                "children": Array [
                  Object {
                    "attributes": Object {
                      "value": ", Hanoi",
                    },
                    "children": Array [],
                    "name": "text",
                  },
                ],
                "name": "inline",
              },
            ],
            "image": null,
          },
        ]
      }
      columnParameters={
        Object {
          "columnCount": 3,
          "columnHeight": 2000,
          "columnLineHeight": 20,
          "columnWidth": 110,
        }
      }
    />
    <View
      style={
        Object {
          "width": 110,
        }
      }
    >
      <MeasureContent
        content={
          Object {
            "children": Array [
              Object {
                "attributes": Object {
                  "value": "line1",
                },
                "children": Array [],
                "name": "text",
              },
            ],
            "id": "p1",
            "name": "paragraph",
          }
        }
        style={
          Object {
            "fontSize": 14,
            "lineHeight": 18,
          }
        }
      />
    </View>
  </View>
</RCTScrollView>
`;
